File: src/python/multiprocessing/11_process_semaphore.py
"""
This program demonstrates the use of a semaphore with multiprocessing in Python.
A semaphore is a synchronization primitive that can be used to control access to
a shared resource by multiple processes. Semaphores are often used to limit the
number of processes that can access a resource at the same time. This can be
useful for managing access to resources with limited capacity, such as
network connections or shared memory.

In this example, we create a pool of worker processes and use a semaphore to
limit the number of processes that can access a shared resource (in this case,
printing to the console) at the same time. The worker processes acquire the
semaphore before accessing the shared resource and release it when they are
finished. This ensures that the maximum number of concurrent accesses to the
shared resource is limited by the semaphore's initial value.
"""

import multiprocessing
import time
import random

# The shared resource is printing to the console
def shared_resource(process_id):
    print(f"Process {process_id} is using the shared resource")
    time.sleep(random.uniform(0.5, 1))
    print(f"Process {process_id} is done using the shared resource")


def worker(semaphore, process_id):
    print(f"Process {process_id} is waiting for the semaphore")
    with semaphore:
        print(f"Process {process_id} acquired the semaphore")
        shared_resource(process_id)
        print(f"Process {process_id} released the semaphore")


def main():
    num_processes = 10
    max_concurrent = 3

    # Create a semaphore with an initial value of max_concurrent
    semaphore = multiprocessing.Semaphore(max_concurrent)

    # Create a pool of worker processes
    with multiprocessing.Pool(processes=num_processes) as pool:
        # Execute the worker function with the semaphore and process IDs
        pool.starmap(worker, [(semaphore, i) for i in range(num_processes)])


if __name__ == "__main__":
    main()

File: src/cpp/multiprocessing/11_process_semaphore.cpp
#include <chrono>
#include <fcntl.h>
#include <iostream>
#include <random>
#include <semaphore.h>
#include <sys/wait.h>
#include <thread>
#include <unistd.h>
#include <vector>

// The shared resource is printing to the console
void shared_resource(int process_id) {
  std::cout << "Process " << process_id << " is using the shared resource"
            << std::endl;
  std::this_thread::sleep_for(std::chrono::milliseconds(rand() % 500 + 500));
  std::cout << "Process " << process_id << " is done using the shared resource"
            << std::endl;
}

void worker(sem_t *semaphore, int process_id) {
  std::cout << "Process " << process_id << " is waiting for the semaphore"
            << std::endl;
  sem_wait(semaphore);
  std::cout << "Process " << process_id << " acquired the semaphore"
            << std::endl;
  shared_resource(process_id);
  std::cout << "Process " << process_id << " released the semaphore"
            << std::endl;
  sem_post(semaphore);
}

int main() {
  const int num_processes = 10;
  const int max_concurrent = 3;

  sem_t *semaphore = sem_open("/semaphore_example", O_CREAT, S_IRUSR | S_IWUSR,
                              max_concurrent);

  std::vector<pid_t> children;

  for (int i = 0; i < num_processes; ++i) {
    pid_t pid = fork();

    if (pid == 0) { // child process
      worker(semaphore, i);
      exit(0);
    } else if (pid > 0) { // parent process
      children.push_back(pid);
    } else {
      std::cerr << "Fork failed" << std::endl;
      return 1;
    }
  }

  // Wait for all child processes to finish
  for (const auto &child_pid : children) {
    waitpid(child_pid, nullptr, 0);
  }

  sem_close(semaphore);
  sem_unlink("/semaphore_example");

  return 0;
}
